src/main/java/org/elasticsearch/index/gateway/IndexShardGateway.java

```diff
@@ -166,113 +144,16 @@ public void recover(boolean indexShouldExists, RecoveryState recoveryState) thro
                 logger.debug("failed to list file details", e);
             }
 
-            Path recoveringTranslogFile = null;
-            if (translogId == -1) {
-                logger.trace("no translog id set (indexShouldExist [{}])", indexShouldExists);
-            } else {
-                final Translog translog = indexShard.translog();
-                final Path translogName = translog.getPath(translogId);
-                logger.trace("try recover from translog file {} locations: {}", translogName, Arrays.toString(translog.locations()));
-                OUTER:
-                for (Path translogLocation : translog.locations()) {
-                    // we have to support .recovering since it's a leftover from previous version but might still be on the filesystem
-                    // we used to rename the foo into foo.recovering since foo was reused / overwritten but we fixed that in 1.5
-                    for (Path recoveryFiles : FileSystemUtils.files(translogLocation, translogName.getFileName() + "{.recovering,}")) {
-                        logger.trace("Translog file found in {}", recoveryFiles);
-                        recoveringTranslogFile = recoveryFiles;
-                        break OUTER;
-                    }
-                    logger.trace("Translog file NOT found in {} - continue", translogLocation);
-                }
-            }
             // we must do this *after* we capture translog name so the engine creation will not make a new one.
             // also we have to do this regardless of whether we have a translog, to follow the recovery stages.
-            indexShard.prepareForTranslogRecovery();
-
-            if (recoveringTranslogFile == null || Files.exists(recoveringTranslogFile) == false) {
-                // no translog files, bail
-                recoveryState.getTranslog().totalOperations(0);
-                recoveryState.getTranslog().totalOperationsOnStart(0);
-                indexShard.finalizeRecovery();
-                indexShard.postRecovery("post recovery from gateway, no translog");
-                // no index, just start the shard and bail
-                return;
-            }
+            final Engine.TranslogRecoveryStats translogRecoveryStats = indexShard.performTranslogRecovery();
+            recoveryState.getTranslog().addRecoveredOperations(translogRecoveryStats.recoveredOperations);
```

Comment:
I think this is tricky for gateway recovery because it will report all the recovered operations at once and not as it goes.  I The `TranslogRecoveryPerformer` can easily have access to the RecvoeryState (it's on the IndexShard). I think it will be better if we increment it directly there.